shared/domains/chelonia/chelonia.js

```diff
@@ -829,30 +831,199 @@ export default (sbp('sbp/selectors/register', {
   // TODO: r.body is a stream.Transform, should we use a callback to process
   //       the events one-by-one instead of converting to giant json object?
   //       however, note if we do that they would be processed in reverse...
-  'chelonia/out/eventsAfter': async function (contractID: string, since: string) {
-    const events = await fetch(`${this.config.connectionURL}/eventsAfter/${contractID}/${since}`, { signal: this.abortController.signal })
-      .then(handleFetchResult('json'))
-    if (Array.isArray(events)) {
-      return events.reverse().map(b64ToStr)
+  'chelonia/out/eventsAfter': function (contractID: string, since: string, limit?: number) {
+    const fetchEventsStreamReader = async () => {
+      requestLimit = Math.min(limit ?? MAX_EVENTS_AFTER, remainingEvents)
+      const eventsResponse = await fetch(`${this.config.connectionURL}/eventsAfter/${contractID}/${since}/${requestLimit}`, { signal })
+      if (!eventsResponse.ok) throw new Error('Unexpected status code')
+      if (!eventsResponse.body) throw new Error('Missing body')
+      latestHEAD = eventsResponse.headers.get('shelter-headinfo-head')
```

Comment:
Should we check the type on this also?

Reply:
We could, but if it's not set or incorrect, it should fail later here:

```
if (remainingEvents === 0 || since === latestHEAD) {
```